src/codecs/xml/index.ts

```diff
@@ -0,0 +1,167 @@
+   *
+   * @param file The `VFile` to decode
+   * @returns A promise that resolves to a Stencila `Node`
+   */
+  public readonly decode = async (
+    file: vfile.VFile
+  ): Promise<stencila.Node> => {
+    const content = await vfile.dump(file)
+    const doc = xml.load(content) as xml.Element
+    const node = decodeDoc(doc)
+    if (node === undefined) {
+      log.warn(`Unable to parse content as Stencila XML`)
+      return null
+    }
+    return node
+  }
+
+  /**
+   * Encode a Stencila `Node` to a `VFile` with XML content.
+   *
+   * @param thing The Stencila `Node` to encode
+   * @returns A promise that resolves to a `VFile`
+   */
+  public readonly encode = async (
+    node: stencila.Node,
+    options: GlobalEncodeOptions = this.defaultEncodeOptions
+  ): Promise<vfile.VFile> => {
+    const doc = await encodeDoc(node, options.isStandalone)
+    const content = xml.dump(doc, { spaces: 4 })
+    return vfile.load(content)
+  }
+}
+
+/**
+ * Decode an XML document to a Stencila `Node`
+ *
+ * @param doc The top level XML element to decode
+ */
+export const decodeDoc = (doc: xml.Element): stencila.Node | undefined => {
+  if (doc !== undefined) {
+    const stencila = xml.first(doc, 'stencila')
+    if (
+      stencila !== null &&
+      stencila.elements !== undefined &&
+      stencila.elements.length === 1
+    ) {
+      const root = stencila.elements[0]
+      return decodeElem(root)
+    }
+  }
+  return undefined
+}
+
+/**
+ * Encode a `Node` as an XML document
+ *
+ * @param node The `Node` to encode
+ * @param standalone Should the document be a standalone XML document?
+ */
+export const encodeDoc = async (
+  node: stencila.Node,
+  standalone: boolean = false
+): Promise<xml.Element> => {
+  const root = encodeNode(node)
+
+  const schemaVersion = await getSchemaVersion('minor')
+  const stencila = xml.elem(
+    'stencila',
+    { xmlns: `https://schema.stenci.la/${schemaVersion}` },
+    root
+  )
+
+  return {
+    elements: [stencila],
+    ...(standalone
+      ? {
+          declaration: {
+            attributes: {
+              version: '1.0',
+              encoding: 'utf-8'
+            }
+          }
+        }
+      : {})
+  }
+}
+
+/**
+ * Decode an XML element to a Stencila `Node`
+ *
+ * @param elem The XML element to decode
+ */
+export const decodeElem = (elem: xml.Element): stencila.Node => {
+  const { name, elements = [] } = elem
+  switch (name) {
+    case 'null':
+      return null
+    case 'boolean':
+      return xml.text(elem) === 'true'
+    case 'number':
+      return parseFloat(xml.text(elem))
+    case 'string':
+      return xml.text(elem)
+    case 'array':
+      return elements.map(decodeElem)
+    default:
+      const node = elements
+        .map(child => {
+          const key = xml.attr(child, 'key') || ''
+          return { [key]: decodeElem(child) }
+        })
+        .reduce((prev, curr) => ({ ...prev, ...curr }), {})
```

Comment:
Are the `keys unique within a collection of elements?
Also, instead of iterating over the collection twice we can combine the `map` operation within the `reduce`.
Something like:

```ts
.reduce((prev, curr) => {
    const key = xml.attr(curr, 'key') || ''
    return {
        ...prev,
        [key]: decodeElem(curr)
    }
}, {})
```

Reply:
Yes, the `keys` are unique for this encoding - just as the keys of a JSON object are unique. There are no guarantees of course that some externally generated XML won't have duplicate keys - but I don't think we need to concern ourselves with that malformed content. Thanks for the suggested change, will make it.